### PR TITLE
[rllib] Port IMPALA to ModelV2/build_tf_policy

### LINKED ISSUE #5133: [rllib] ModelV2 breaks APPO+A2C / Atari
Tracking issue from https://github.com/ray-project/ray/pull/5130

The symptom is that entropy collapses to zero immediately at the start of training.

### PR DESCRIPTION
**Pull request by ericl**
This also considerably simplifies APPOPolicy as it can be derived from IMPALA with just a couple changes.

Closes https://github.com/ray-project/ray/issues/5133

### CODE REVIEW

File: python/ray/rllib/agents/ppo/appo_policy.py [hunk 1]
```diff
@@ -337,61 +276,8 @@ def add_values_and_logits(policy):
...
-def setup_mixins(policy, obs_space, action_space, config):
-    LearningRateSchedule.__init__(policy, config["lr"], config["lr_schedule"])
-    ValueNetworkMixin.__init__(policy)
-
-
-AsyncPPOTFPolicy = build_tf_policy(
-    name="AsyncPPOTFPolicy",
-    get_default_config=lambda: ray.rllib.agents.impala.impala.DEFAULT_CONFIG,
+AsyncPPOTFPolicy = VTraceTFPolicy.with_updates(
+    name="VTraceTFPolicy",
```
Comment (pcmoritz) [inline python/ray/rllib/agents/ppo/appo_policy.py]: should this be "AsyncPPOTFPolicy"?

File: python/ray/rllib/agents/ppo/appo_policy.py [hunk 2]
```diff
@@ -163,41 +159,6 @@ def __init__(self,
                            self.entropy * entropy_coeff)
 
 
-def _make_time_major(policy, tensor, drop_last=False):
```
Comment (pcmoritz) [inline python/ray/rllib/agents/ppo/appo_policy.py]: It looks like the functions in this file was moved to vtrace_policy.py without change, is that right?

File: python/ray/rllib/agents/impala/vtrace_policy.py [hunk 3]
```diff
@@ -107,237 +107,190 @@ def __init__(self,
...
-            mask = tf.sequence_mask(self.model.seq_lens, max_seq_len)
-            mask = tf.reshape(mask, [-1])
-        else:
-            mask = tf.ones_like(rewards, dtype=tf.bool)
-
-        # Prepare actions for loss
-        loss_actions = actions if is_multidiscrete else tf.expand_dims(
-            actions, axis=1)
-
-        # Inputs are reshaped from [B * T] => [T - 1, B] for V-trace calc.
```
Comment (pcmoritz) [inline python/ray/rllib/agents/impala/vtrace_policy.py]: consider retaining this comment

File: python/ray/rllib/agents/impala/vtrace_policy.py [hunk 4]
```diff
@@ -107,237 +107,190 @@ def __init__(self,
...
-            (SampleBatch.DONES, dones),
-            (BEHAVIOUR_LOGITS, behaviour_logits),
-            (SampleBatch.REWARDS, rewards),
-            (SampleBatch.CUR_OBS, observations),
-            (SampleBatch.PREV_ACTIONS, prev_actions),
-            (SampleBatch.PREV_REWARDS, prev_rewards),
-        ]
-        LearningRateSchedule.__init__(self, self.config["lr"],
-                                      self.config["lr_schedule"])
-        TFPolicy.__init__(
```
Comment (pcmoritz) [inline python/ray/rllib/agents/impala/vtrace_policy.py]: this was replaced with a call to `build_tf_policy`

File: python/ray/rllib/agents/impala/vtrace_policy.py [hunk 5]
```diff
@@ -107,237 +107,190 @@ def __init__(self,
...
+
+
+def stats(policy, batch_tensors):
+    values_batched = _make_time_major(
+        policy, policy.value_function, drop_last=policy.config["vtrace"])
+
+    return {
+        "cur_lr": tf.cast(policy.cur_lr, tf.float64),
+        "policy_loss": policy.loss.pi_loss,
+        "entropy": policy.loss.entropy,
```
Comment (pcmoritz) [inline python/ray/rllib/agents/impala/vtrace_policy.py]: Was dropping `grad_gnorm` intentional? Should probably be kept!

Comment (pcmoritz) [reply python/ray/rllib/agents/impala/vtrace_policy.py]: ah nvm, it was moved to `grad_stats`

**Review by pcmoritz — APPROVED** [review]
LGTM! Does it produce identical numerical results to the old implementation (or if not, the same performance?)

File: python/ray/rllib/agents/ppo/appo_policy.py (hunk 2 above)
Comment (ericl) [reply python/ray/rllib/agents/ppo/appo_policy.py]: Yep

**Comment by ericl** [comment]
Sadly, this seems to break learning on Breakout (stuck at 2 reward after 1m timesteps).

**Comment by ericl** [comment]
The failure mode is quite strange, it forces entropy to 0 very quickly.
